src/stitcher/image_loader.rs

```diff
@@ -1,26 +1,65 @@
 //! This module is for all methods involved in getting selected images loaded into memory.
 
-use anyhow::anyhow;
 use image::{
-    image_dimensions, imageops::FilterType::Lanczos3, GenericImage, ImageReader, RgbImage,
+    error::ImageError, image_dimensions, imageops::FilterType::Lanczos3, GenericImage, ImageReader,
+    RgbImage,
 };
 use rayon::iter::{IntoParallelIterator, ParallelIterator};
-use std::{fs::read_dir, path::PathBuf, time::Instant};
+use std::{
+    fs::read_dir,
+    path::{Path, PathBuf},
+    time::Instant,
+};
 use thiserror::Error;
 
 #[derive(Error, Debug)]
+/// Errors related to loading images.
 pub enum ImageLoaderError {
+    #[error("Expected a directory at \"{0}\"")]
+    ExpectedDirectory(PathBuf),
+    #[error("Could not find a valid path at \"{0}\"")]
+    NotFound(PathBuf),
+    #[error("Permission denied while attempting to access \"{0}\"")]
+    PermissionDenied(PathBuf),
     #[error("No images were found in the selected directory")]
     NoImagesInDirectory,
+    // TODO: convert upstream errors to more specific errors
+    #[error("{0}")]
+    ImageError(ImageError),
+}
+
+impl From<ImageError> for ImageLoaderError {
+    fn from(value: ImageError) -> Self {
+        Self::ImageError(value)
+    }
+}
+
+impl ImageLoaderError {
+    fn from_io_error(err: std::io::Error, path: PathBuf) -> ImageLoaderError {
+        use std::io::ErrorKind as Kind;
+        match err.kind() {
+            Kind::NotFound => ImageLoaderError::NotFound(path),
+            Kind::PermissionDenied => ImageLoaderError::PermissionDenied(path),
+            _ => unimplemented!(),
+        }
+    }
 }
 
 /// Finds all `.jpg`, `.jpeg`, `.png` and `.webp` images within a directory.
 ///
 /// Throws an error if:
 ///  - The directory is invalid or does not contain any images.
 ///  - The directory does not contain any jpg, jpeg, png, or webp images.
-pub fn find_images(directory_path: &str) -> anyhow::Result<Vec<PathBuf>> {
-    let mut images: Vec<_> = read_dir(directory_path)?
+pub fn find_images(directory_path: impl AsRef<Path>) -> Result<Vec<PathBuf>, ImageLoaderError> {
```

Comment:
Changed this function to accept a generic parameter to make it more flexible. This means you can, for instance, provide a `&PathBuf` instead of a `&str` (useful for application code). 

Note that there *is* a compile-time cost due to monomorphization, but if that becomes a problem, I can fix it with the [non-generic inner function hack](https://www.possiblerust.com/pattern/non-generic-inner-functions)